Parser semicolon in intializer expression error handling

ArcadeMode opened the pull request:
Suggested fix for #22790 
Currently only implemented for List, Object and Array initializers and not for other list expressions such as e.g. attribute lists or base lists.

cston commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -12914,8 +12914,27 @@ private PostSkipAction SkipBadInitializerListTokens<T>(ref SyntaxToken startToke
+                abortFunc,
                 expected);
+
+            static bool abortFunc(LanguageParser p)
```
>abortFunc [](http://example.com/codeflow?start=24&length=9)

The `abortFunc()` local function looks identical here and in `SkipBadArrayInitializerTokens()`. Consider creating a single helper method to reuse across the two callers.

cston commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -12914,8 +12914,27 @@ private PostSkipAction SkipBadInitializerListTokens<T>(ref SyntaxToken startToke
+            }
+        }
+
+        private T IsNextToken<T>(Func<LanguageParser, T> nextTokenParseFunc)
```
>T [](http://example.com/codeflow?start=54&length=1)

`bool`?

cston commented:
It looks like there are several test failures related to parsing.

cston commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs:
```diff
@@ -4925,14 +4925,43 @@
-            Assert.Equal(1, file.Errors().Length);
-            Assert.Equal((int)ErrorCode.ERR_RbraceExpected, file.Errors()[0].Code);
+            var errors = file.Errors();
+            Assert.Equal(3, errors.Length);
```
>Assert.Equal(3, errors.Length); [](http://example.com/codeflow?start=12&length=31)

We appear to be reporting more errors in several of these cases. Is that expected?

ArcadeMode replied:
I do consider this expected. Before the `;` was interpreted as the end of the initializer expression (as if a `}` was written). With these changes, if a `}` or expression follows the semicolon, the `;` will not terminate the initializer expression. 

As a result more accurate errors are generated (imo). In this test case for example the 'new' errors are 1. a syntax error as a `;` is not expected in an initializer, 2. a missing `;` after the next `}` (which terminates the initializer expression) and 3. the missing `}` to close the outer-most scope.

ArcadeMode commented:
resolved feedback, still seeing failed checks but afaik is unrelated to the changes in this PR.

RikkiGibson commented:
I think we would like to see this improvement introduced in any "list-like" syntax which is wrapped in brackets. Switch expressions, for example. `obj switch { Pattern1 => a; Pattern2 => b }` has similarly bad error recovery as object initializers. Thanks @333fred for pointing this out.

If you are interested, you could consider opening a new PR after this one to also handle that case, or expand this PR to also handle it if you think that is manageable. @cston let me know if this advice seems right to you.

cston commented:
@ArcadeMode, after discussing with several team members, we thought we should have tests for motivating scenarios, to capture the updated behavior and avoid regressions in the future.

For instance, perhaps the following tests which have much better error recovery (and fewer parsing errors and no binding errors) with this change (thanks @333fred for the scenario). Thanks.

```csharp
[Fact]
public void SemicolonSeparatorInArrayInitializer()
{
    var source = """
        class Program
        {
            static void Main()
            {
                F();
                var a = new[] { 1; 2 };
            }
            static object F()
            {
                return null;
            }
        }
        """;
    var comp = CreateCompilation(source);
    comp.VerifyDiagnostics(
        // (6,26): error CS1003: Syntax error, ',' expected
        //         var a = new[] { 1; 2 };
        Diagnostic(ErrorCode.ERR_SyntaxError, ";").WithArguments(",").WithLocation(6, 26),
        // (6,28): error CS1003: Syntax error, ',' expected
        //         var a = new[] { 1; 2 };
        Diagnostic(ErrorCode.ERR_SyntaxError, "2").WithArguments(",").WithLocation(6, 28));
}

[Fact]
public void SemicolonSeparatorInObjectInitializer()
{
    var source = """
        class Program
        {
            static void Main()
            {
                F();
                var o = new { A = 1; B = 2 };
            }
            static object F()
            {
                return null;
            }
        }
        """;
    var comp = CreateCompilation(source);
    comp.VerifyDiagnostics(
        // (6,28): error CS1003: Syntax error, ',' expected
        //         var o = new { A = 1; B = 2 };
        Diagnostic(ErrorCode.ERR_SyntaxError, ";").WithArguments(",").WithLocation(6, 28),
        // (6,30): error CS1003: Syntax error, ',' expected
        //         var o = new { A = 1; B = 2 };
        Diagnostic(ErrorCode.ERR_SyntaxError, "B").WithArguments(",").WithLocation(6, 30));
}
```

cston commented:
> I think we would like to see this improvement introduced in any "list-like" syntax which is wrapped in curly braces. Switch expressions, for example. ...

Agreed, and as you suggested @RikkiGibson, that could be handled in a separate PR.

333fred commented:
It would also be good to have scenarios that put the `;` on the last element, not the second element.

ArcadeMode commented:
I've added the motivating test scenarios, including tests with the `;` on the last element as suggested.

> I think we would like to see this improvement introduced in any "list-like" syntax which is wrapped in curly braces. Switch expressions, for example. `obj switch { Pattern1 => a; Pattern2 => b }` has similarly bad error recovery as object initializers. Thanks @333fred for pointing this out.

This sounds like something I'd like to pick up, could we outline which exact syntaxes fall within the above definition? Perhaps make a new issue for it or comment it in #22790?

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -12912,12 +12912,32 @@ private ExpressionSyntax ParseObjectOrCollectionInitializerMember(ref bool isObj
+            {
+                var resetPoint = GetResetPoint();
+                EatToken();
+                bool result = nextTokenParseFunc(this);
```
i don't see the benefit in passing a callback here.  we know exactly the code to run to check this.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -12912,12 +12912,32 @@ private ExpressionSyntax ParseObjectOrCollectionInitializerMember(ref bool isObj
+        {
+            if (!openBraceIsMissing && CurrentToken.Kind == SyntaxKind.SemicolonToken)
+            {
+                return !isNextToken(lp => lp.CurrentToken.Kind == SyntaxKind.CloseBraceToken || lp.IsPossibleVariableInitializer());
```
not really understanding teh reason to call IsPossibleVariableInitializer.  Need docs.

CyrusNajmabadi replied:
i don't get why we're passing the language parser as an argument here.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/Formatting/Indentation/SmartTokenFormatterFormatRangeTests.cs:
```diff
@@ -2855,21 +2855,22 @@ public async Task NoLineChangeWithSyntaxError()
                 expected,
+                SyntaxKind.CommaToken,
                 SyntaxKind.OpenBraceToken);
         }
```
please don't change existing test start-contents.  Just show what teh new behavior is.  Add new tests with the new start-contents.

ArcadeMode commented:
I refactored the method to more clearly describe its purpose. Also added a few comments to explain my intiution.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -12912,12 +12912,35 @@ private ExpressionSyntax ParseObjectOrCollectionInitializerMember(ref bool isObj
+                // 2. another variable intializer - not the end yet
+                var resetPoint = GetResetPoint();
+                EatToken();
+                bool initializerContinues = CurrentToken.Kind == SyntaxKind.CloseBraceToken || IsPossibleVariableInitializer();
```
Note: this looks like it woudl barf on `new Dict { [0] = "a"; [1] = "B" }`.   i'd prefer us be more intentional in what we're looking for as the potential start of a next initializer.

ArcadeMode replied:
added tests that show it doesnt barf now

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -12912,12 +12912,35 @@ private ExpressionSyntax ParseObjectOrCollectionInitializerMember(ref bool isObj
+            // the current token ends the initializer when it is:
+            // 1. a close brace
+            // 2. another type of terminator
+            return CurrentToken.Kind == SyntaxKind.CloseBraceToken || IsTerminator();
```
nit: this is the trivial logic.  my preference with code would be to invert htings so we have:

```c#
if (openBraceIsMissing || CurrentToken.Kind != SyntaxKind.SemicolonToken)
    return trivialLogic;
    
doComplexLogic();
```

it just reads more easily.  thanks!

ArcadeMode replied:
✔️

CyrusNajmabadi commented on src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs:
```diff
@@ -2826,14 +2826,14 @@ static void Main(string[] args)
-        var x = 1/*</bind>*/;
-    }
+        var x = 1;
+    }/*</bind>*/
```
i recommend not changing existing test inputs.  It means we can't tell if the PR affected the output of some scenario.  Instead, keep the old test and add a new test with a new input.  THanks!

ArcadeMode replied:
Sadly I was unable to keep the original input in this case. Since the expression is now read beyond the `;` till the next `}` the test suite would start throwing exceptions. My understanding is that it couldnt find the expression ending with `</bind>` as the expression now ends at the next `}` I moved the `</bind>` there.

CyrusNajmabadi commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs:
```diff
@@ -4480,8 +4482,10 @@ public void TestSemicolonAfterAnonymousTypeMemberStart()
+            Assert.Equal(3, file.Errors().Length);
+            Assert.Equal((int)ErrorCode.ERR_SyntaxError, file.Errors()[0].Code);
+            Assert.Equal((int)ErrorCode.ERR_SemicolonExpected, file.Errors()[1].Code);
+            Assert.Equal((int)ErrorCode.ERR_RbraceExpected, file.Errors()[2].Code);
```
this looks to have gotten worse. :-/

ArcadeMode replied:
(copied my response from another comment)
> I do consider this expected. Before the ; was interpreted as the end of the initializer expression (as if a } was written). With these changes, if a } or expression follows the semicolon, the ; will not terminate the initializer expression.

> As a result more accurate errors are generated (imo). In this test case for example the 'new' errors are 1. a syntax error as a ; is not expected in an initializer, 2. a missing ; after the next } (which terminates the initializer expression) and 3. the missing } to close the outer-most scope.

I would argue they are more accurate than before, what would you say regarding the above reasoning?

CyrusNajmabadi commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs:
```diff
@@ -4562,8 +4570,124 @@ public void TestSemicolonAfterAnonymousTypeMemberComma()
+            Assert.Equal(3, file.Errors().Length);
+            Assert.Equal((int)ErrorCode.ERR_IdentifierExpected, file.Errors()[0].Code);
+            Assert.Equal((int)ErrorCode.ERR_SemicolonExpected, file.Errors()[1].Code);
+            Assert.Equal((int)ErrorCode.ERR_RbraceExpected, file.Errors()[2].Code);
```
these all seem like unfortunate regressions.  this is def a tricky space.  i'm curious if we can do better with detecting if what comes should be eaten into the initializer, or should keep the old logic.

CyrusNajmabadi commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs:
```diff
@@ -4562,8 +4570,124 @@ public void TestSemicolonAfterAnonymousTypeMemberComma()
+                Diagnostic(ErrorCode.ERR_SyntaxError, ";").WithArguments(",").WithLocation(6, 26),
+                // (6,28): error CS1003: Syntax error, ',' expected
+                //         var a = new[] { 1; 2 };
+                Diagnostic(ErrorCode.ERR_SyntaxError, "2").WithArguments(",").WithLocation(6, 28));
```
this doesn't seem correct.  if we are treating `;` as the separator, we should def get the first error, but not the second.  the '2' should be fine now as it's an acceptable expression in the initializer.

We should also dump the tree here (see the UsingTree function that lots of tests use).  So we can show that the error recovery worked.

CyrusNajmabadi replied:
this is currently a blocking concern for me as it indicates something is not working here as i would expect based on the goals of this PR.

ArcadeMode replied:
I've tried a different approach and managed to achieve that the second error does not show up. I'll push the commits hopefully tomorrow. Then the changes should be more in line with what you would expect.

CyrusNajmabadi reviewed (changes requested):
Looking at the tests, things seem off here.  Holding off on the review till this is discussed and understood better.

ArcadeMode commented:
I've implemented a different approach that yields better results (imo). Will wait for feedback

Addressing @RikkiGibson's comment regarding switch expressions, I have added test cases that show that switch patterns dont have as bad error handling as object/array initializers, at least for the tested cases. During switch expression parsing semicolons are eaten as commas, so there is no 'looking ahead' if it should be interpreted as a comma or terminator. This does appear to be very different from object/array initializer parsing.

RikkiGibson commented:
Looks like this slipped through the cracks over the holidays :) Will try and give another review when I have the chance.

RikkiGibson commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -12918,6 +12925,31 @@ private PostSkipAction SkipBadInitializerListTokens<T>(ref SyntaxToken startToke
                 expected);
         }
 
+        private bool IsValidInitializerToken(ref SyntaxToken startToken, Func<bool> isInitializerMemberFn)
```
I'm having a little difficulty figuring out the purpose of everything in this helper. It seems odd to me that we need to use a separate reset point in here. I'm also concerned that we're introducing a capture of `this` in the delegate argument. If we decide to keep this factoring it will be better to have `Func<LanguageParser, bool>` which takes `this` as an argument, and static lambdas at the call site.

I am wondering if the implementation of switch expression arm parsing would be helpful to review as we consider how to factor this.
https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs#L624

CyrusNajmabadi commented on src/Compilers/CSharp/Test/Syntax/Parsing/ParsingErrorRecoveryTests.cs:
```diff
@@ -4480,8 +4484,10 @@ public void TestSemicolonAfterAnonymousTypeMemberStart()
             Assert.Equal(SyntaxKind.AnonymousObjectCreationExpression, ds.Declaration.Variables[0].Initializer.Value.Kind());
-            Assert.Equal(1, file.Errors().Length);
-            Assert.Equal((int)ErrorCode.ERR_RbraceExpected, file.Errors()[0].Code);
+            Assert.Equal(3, file.Errors().Length);
```
that makes sense to me.

RikkiGibson commented:
The parser now has a helper `ParseCommaSeparatedSyntaxList` which takes `bool allowSemicolonAsSeparator` (for error recovery). Making a different decision for error recovery now seems to be a matter of just changing the flag passed at certain call sites and updating the test baselines.

Please let us know if there's any way we can assist with this PR. Feel free to hop in our [discord](https://discord.gg/csharp) or wherever else if that would help. I will mark as draft for now since it is blocked on merging in the new changes from main.

RikkiGibson commented:
I would open a new PR, carrying over the tests from this PR. Thanks!
